### ISSUE TITLE
Errors in Generated Types.go file

### ISSUE THREAD

**Issue by c-reeder** [goadesign/goa#3120]
In my goa design file I have the following type definition:
```
var Workspace = Type("workspace", func() {
	Attribute("id", String, "Workspace ID")
	Attribute("name", String)
	Attribute("description", String)
	Attribute("destination_ids", ArrayOf(String))
	Attribute("logo_url", String)
})
```
and in the same file, I use it as such:
```
	Method("update", func() {
		Description("Update an existing workspace information")
		Payload(Workspace, func() {
			Required("id")
		})

		HTTP(func() {
			PATCH("/{id}")

			Response("workspace_not_found", StatusNotFound)
		})
	})
```
The goa gen command runs successfully without any error output.
Upon trying to compile/run my app, I get the following error in my `gen/http/workspaces/server/types.go` file:
`cannot use id (variable of type string) as type *string in assignment`
Pointed at the following block of generated code:
```
// NewUpdateWorkspace builds a workspaces service update endpoint payload.
func NewUpdateWorkspace(body *UpdateRequestBody, id string) *workspaces.Workspace {
	v := &workspaces.Workspace{
		Name:        body.Name,
		Description: body.Description,
		LogoURL:     body.LogoURL,
	}
	if body.DestinationIds != nil {
		v.DestinationIds = make([]string, len(body.DestinationIds))
		for i, val := range body.DestinationIds {
			v.DestinationIds[i] = val
		}
	}
	v.ID = id

	return v
}
```
where the second to last line `v.ID = id` is the one throwing the error.

Upon quick inspection, I find that the generated `workspaces.Workspace` type above has the `ID` field as a `*string`, but as you can see, the incoming parameter on this method is a non-pointer `string`. 

If I alter my design code above to remove the `Required("id")`line from the method Payload and put it inside the type definition...the error doesn't occur. Similarly if I have the required line in BOTH the Payload definition and the Type definition...no error.

I can see in the [FAQ page](https://goa.design/learn/faq/) the reasons why generated fields are sometimes pointers and sometimes not, however the mismatch allowed to result in the generated code is what I believe is an issue. 

So that leads me to the following questions:

1. If I'm not supposed to do the above, why doesn't the goa gen command give me an error to let me know rather than generating non-functioning code?
2. If I'm not supposed to do the above, then why is there an example just like it in the dsl documentation? If you look at the `Payload()` [examples](https://pkg.go.dev/goa.design/goa/dsl#Payload), the last example "divide", clearly is defining making attributes required on a user defined Type from the Payload method (same as me). 

Upon seeing that example in the dsl docs, it makes sense to me that it would allow that, because you may want to take the same Type in the Payload for several methods, but you only need to make a certain field in it required for some but not all. Otherwise, you'd have to have duplicates of the Type (either inlined or separately defined) with different Required calls. 

I'm really loving goa so far and I'm hoping we can find a solution to this!

Thanks in advance for your help!

**Comment by Arduinoid** [comment]
I can confirm this behavior as well. I would love to know why this happens and what the solution should be. Whether it's a fix or correcting DSL expectations

**Comment by raphael** [comment]
I am not able to reproduce this, using the latest Goa the `ID` field gets generated as a `string` and not `*string`:
```go
// Workspace is the payload type of the workspaces service update method.
type Workspace struct {
	// Workspace ID
	ID             string
	Name           *string
	Description    *string
	DestinationIds []string
	LogoURL        *string
}
```
It could be because of something else in the design. Could you provide a complete repro?

**Comment by c-reeder** [comment]
@raphael , sorry for the delay, but I have created an example repo demonstrating part of the issue that is described above. 
[goa_issue](https://github.com/c-reeder/goa_issue)
Feel free to explore the code, but it's really just the `design/design.go` file and the `main.go` file.
As I put in the readme, just run the following:
```
make goa-gen
make build
```
You will see that the build fails with the same error I mentioned.

I used the same example from above, but to expand on my description...here is what I have observed.

The issue seems to be that when you have a user defined type.....and then you have inline changes like these:
```
		Payload(Workspace, func() {
			Required("id")
		})
```
...the generator will combine the separate type definition with these inline additions to generate a single type which it attempts to use for all methods that use that type. However that fails because it generates types that don't match each other.

I'm not sure that the issue is limited to this. Maybe @Arduinoid can chime in with anything further he has observed.

I hope this example better helps us answer my original questions above.

Thanks again!
